subsys/bluetooth/services/dis.c

```diff
@@ -45,105 +47,231 @@ static struct dis_pnp dis_pnp_id = {
+#define BT_DIS_SERIAL_NUMBER_STR_REF CONFIG_BT_DIS_SERIAL_NUMBER_STR
+#define BT_DIS_FW_REV_STR_REF        CONFIG_BT_DIS_FW_REV_STR
+#define BT_DIS_HW_REV_STR_REF        CONFIG_BT_DIS_HW_REV_STR
+#define BT_DIS_SW_REV_STR_REF        CONFIG_BT_DIS_SW_REV_STR
+#define BT_DIS_UDI_LABEL_STR_REF     CONFIG_BT_DIS_UDI_LABEL_STR
+#define BT_DIS_UDI_DI_STR_REF        CONFIG_BT_DIS_UDI_DI_STR
+#define BT_DIS_UDI_ISSUER_STR_REF    CONFIG_BT_DIS_UDI_ISSUER_STR
+#define BT_DIS_UDI_AUTHORITY_STR_REF CONFIG_BT_DIS_UDI_AUTHORITY_STR
+#define BT_DIS_IEEE_RCDL_STR_REF     CONFIG_BT_DIS_IEEE_RCDL_STR
+
+#define BT_DIS_STRLEN(x) strlen(x)
 
 #endif /* CONFIG_BT_DIS_SETTINGS */
 
-static ssize_t read_str(struct bt_conn *conn,
-			  const struct bt_gatt_attr *attr, void *buf,
-			  uint16_t len, uint16_t offset)
+static ssize_t read_str(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			uint16_t len, uint16_t offset)
 {
 	return bt_gatt_attr_read(conn, attr, buf, len, offset, attr->user_data,
-				 strlen(attr->user_data));
+				 BT_DIS_STRLEN(attr->user_data));
 }
 
 #if CONFIG_BT_DIS_PNP
-static ssize_t read_pnp_id(struct bt_conn *conn,
-			   const struct bt_gatt_attr *attr, void *buf,
+static ssize_t read_pnp_id(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
 			   uint16_t len, uint16_t offset)
 {
-	return bt_gatt_attr_read(conn, attr, buf, len, offset, &dis_pnp_id,
-				 sizeof(dis_pnp_id));
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &dis_pnp_id, sizeof(dis_pnp_id));
+}
+#endif
+
+#if CONFIG_BT_DIS_SYSTEM_ID
+static ssize_t read_system_id(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			      uint16_t len, uint16_t offset)
+{
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, dis_system_id,
+				 sizeof(dis_system_id));
+}
+#endif
+
+#if CONFIG_BT_DIS_UDI
+#define BT_DIS_UDI_FLAG_LABEL     (!!BT_DIS_UDI_LABEL_STR_REF[0])
+#define BT_DIS_UDI_FLAG_DI        (!!BT_DIS_UDI_DI_STR_REF[0])
+#define BT_DIS_UDI_FLAG_ISSUER    (!!BT_DIS_UDI_ISSUER_STR_REF[0])
+#define BT_DIS_UDI_FLAG_AUTHORITY (!!BT_DIS_UDI_AUTHORITY_STR_REF[0])
+#define BT_DIS_UDI_FLAGS                                                                           \
+	(BT_DIS_UDI_FLAG_LABEL | (BT_DIS_UDI_FLAG_DI << 1) | (BT_DIS_UDI_FLAG_ISSUER << 2) |       \
+	 (BT_DIS_UDI_FLAG_AUTHORITY << 3))
+
+static ssize_t read_udi(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			uint16_t len, uint16_t offset)
+{
+	uint16_t written_bytes = 0;
+	uint16_t index = offset;
+
+	const int16_t udi_label_len =
+		BT_DIS_UDI_FLAG_LABEL ? BT_DIS_STRLEN(BT_DIS_UDI_LABEL_STR_REF) : -1;
+	const int16_t udi_di_len = BT_DIS_UDI_FLAG_DI ? BT_DIS_STRLEN(BT_DIS_UDI_DI_STR_REF) : -1;
+	const int16_t udi_issuer_len =
+		BT_DIS_UDI_FLAG_ISSUER ? BT_DIS_STRLEN(BT_DIS_UDI_ISSUER_STR_REF) : -1;
+	const int16_t udi_authority_len =
+		BT_DIS_UDI_FLAG_AUTHORITY ? BT_DIS_STRLEN(BT_DIS_UDI_AUTHORITY_STR_REF) : -1;
```

Comment:
Hmm, this looks needlessly complicated. 

The `BT_DIS_UDI_FLAG_` values are effectively booleans based on the first character in the strings, right, and will 0 for the empty string and 1 for any other string. 

But doesn't that also mean that we could do 
```
const uint8_t udi_label_len = BT_DIS_STRLEN(BT_DIS_UDI_LABEL_STR_REF);
```

which would also have either the value `0` for the empty string and a non-0 value for any other string? 

It should be fairly easy to update the below to allow for these lengths to be `0` instead of `-1` for the empty strings

Reply:
I could for example swap to
```c
	const int16_t udi_label_len = CHECK_ZERO(BT_DIS_STRLEN(BT_DIS_UDI_LABEL_STR_REF));
	const int16_t udi_di_len = CHECK_ZERO(BT_DIS_STRLEN(BT_DIS_UDI_DI_STR_REF));
	const int16_t udi_issuer_len = CHECK_ZERO(BT_DIS_STRLEN(BT_DIS_UDI_ISSUER_STR_REF));
	const int16_t udi_authority_len = CHECK_ZERO(BT_DIS_STRLEN(BT_DIS_UDI_AUTHORITY_STR_REF));
```

where 
```c
#define CHECK_ZERO(x) (x == 0 ? -sizeof('\0') : x)
```

Reply:
Renamed the macro to `SUB_NULLTERM`. Other suggestions to a better name are welcome.

Reply:
But why not just `const uint8_t udi_label_len = BT_DIS_STRLEN(BT_DIS_UDI_LABEL_STR_REF);`? Why does it need to become a negative number when both `-1` and `0` both indicate the empty string?

Reply:
The value `-1` is there so that the null-terminator is not included in the offset calculation. This is needed since the strings not present in flags should not be present at all. I.e. they should not have their null-terminator there.

However, I now see that this could just have been done by defining a macro for the size of the string plus null-terminator if present, like I did for the configs.

```c
#define ADD_NULLTERM(x) (x > 0 ? x + sizeof('\0') : 0)
```

Reply:
Something is getting really messy in this service. 

We can't have strings that maybe/maybe not are NULL terminated an then rely on string functions like `strlen` (`BT_DIS_STRLEN`) which won't work if they aren't NULL terminated. 

The loop here is fairly hard to read with several similarly named variables and many updates per `if` statement. 

It's probably correctly implemented (though with no CI testing it's hard to tell, especially since I'm pretty sure `SUB_NULLTERM(BT_DIS_STRLEN(...))` doesn't work correctly due to the strings possibly not being NULL terminated. 

A suggestion from me would be to do something similar to what @alwa-nordic suggested. 

```c
static size_t read_udi_subval(uint8_t *val, size_t val_size, void *buf, uint16_t buf_size, uint16_t buf_len, uint16_t offset) { ... };

static ssize_t read_udi(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
			uint16_t len, uint16_t offset)
{
	uint16_t bytes_read = 0;	

	bytes_read += read_udi_subval(BT_DIS_UDI_LABEL_STR_REF, BT_DIS_STRLEN(BT_DIS_UDI_LABEL_STR_REF), buf, len, bytes_read, offset;
	bytes_read += read_udi_subval(BT_DIS_UDI_DI_STR_REF, BT_DIS_STRLEN(BT_DIS_UDI_DI_STR_REF), buf, len, bytes_read, offset;
	bytes_read += read_udi_subval(BT_DIS_UDI_ISSUER_STR_REF, BT_DIS_STRLEN(BT_DIS_UDI_ISSUER_STR_REF), buf, len, bytes_read, offset;
	bytes_read += read_udi_subval(BT_DIS_UDI_AUTHORITY_STR_REF, BT_DIS_STRLEN(BT_DIS_UDI_AUTHORITY_STR_REF), buf, len, bytes_read, offset;

	return bytes_read;
}
```

where `read_udi_subval` would skip values not within the `offset + len` value range, and otherwise return the number of bytes read of the specific value. 

This approach may have additional calls to `read_udi_subval` but they would `return 0` early if not being read, e.g. if `BT_DIS_STRLEN(BT_DIS_UDI_LABEL_STR_REF) = 10` and `offset = 11` then it's just a no-op.

Reply:
Regarding the null-terminators: the arrays will always have a null-terminator if it has fewer characters than max. If settings mode is enabled, then `BT_DIS_STRLEN` is `strnlen` with max as `BT_DIS_STR_MAX`. If not, then it is just `strlen`. If settings is disabled, then the string will come from Kconfig, and it is always null-terminated.
If in settings, we will always risk that the null-terminator will be dropped if the string is too long. Therefore I think that using `strnlen` is the best solution. Length is the only thing we need the null-terminator for. The alternative is setting the array[BT_DIS_STR_MAX] = 0 inside read_str.

The substrings could probably be split into different methods, but the flags would need to be handed a way, and we would need to keep track of current index and pass to the methods.

Reply:
This would be my suggestion:
```c
static void read_udi_subval(const char *str, uint16_t str_len, uint8_t *buf, uint16_t *bytes_read,
			    uint16_t *index, uint16_t len, uint16_t offset)
{
	/* String should not be with included null-terminator if empty */
	if (str_len == 0) {
		return;
	}

	for (uint16_t i = 0; i < str_len; i++) {
		if (*index >= offset && *bytes_read < len) {
			buf[*bytes_read] = str[i];

			(*bytes_read)++;
		}

		(*index)++;
	}

	/* Null-terminator */
	if (*index >= offset && *bytes_read < len) {
		buf[*bytes_read] = 0;

		(*bytes_read)++;
	}

	(*index)++;
}

static ssize_t read_udi(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
			uint16_t len, uint16_t offset)
{
	uint16_t bytes_read = 0;

	uint8_t *buf_i = (uint8_t *)buf;

	/* Flag */
	uint16_t index = sizeof(uint8_t);
	if (offset == 0) {
		buf_i[0] = BT_DIS_UDI_FLAGS;
	}

	read_udi_subval(BT_DIS_UDI_LABEL_STR_REF, BT_DIS_STRLEN(BT_DIS_UDI_LABEL_STR_REF), buf_i,
			&bytes_read, &index, len, offset);
	read_udi_subval(BT_DIS_UDI_DI_STR_REF, BT_DIS_STRLEN(BT_DIS_UDI_DI_STR_REF), buf_i,
			&bytes_read, &index, len, offset);
	read_udi_subval(BT_DIS_UDI_ISSUER_STR_REF, BT_DIS_STRLEN(BT_DIS_UDI_ISSUER_STR_REF), buf_i,
			&bytes_read, &index, len, offset);
	read_udi_subval(BT_DIS_UDI_AUTHORITY_STR_REF, BT_DIS_STRLEN(BT_DIS_UDI_AUTHORITY_STR_REF),
			buf_i, &bytes_read, &index, len, offset);

	return bytes_read;
}
```

Also, the length thing, we should make the arrays `CONFIG_BT_DIS_STR_MAX + 1` long, so that one will not get a warning if assigning a string literal of `CONFIG_BT_DIS_STR_MAX`.

Reply:
I think the above design looks good. A few comments to it, but I think that's easier to point out in a proper review

Reply:
I would prefer that we make sure the strings are always stored NULL-terminated in their respective variables. That removes any need for NULL-termination logic in the GATT read callback. It would also get rid of `BT_DIS_STRLEN`.

Reply:
I changed it to `char` now